---
title: "103-100-inds"
output: html_document
date: "2023-09-20"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

```{r}
library(tidyverse)
```

```{r}
stats<-read_csv("outputs/102/alignment-info.txt", col_names=c("Sample","Reads","Coverage"))
sub<-filter(stats, Reads > (325*200))
```


```{r,eval=FALSE}
test<-sub %>% slice_sample(n=100)
write_tsv(test, file="meta/test100.csv")
```


Now we need commands like this

`../scripts/GTseq_Genotyper_v3_Tetra.pl ../genotyper_input.csv i001_10_4_18225.fastq > i001_10_4_18225.genos`

or in my case 
`./data-processing-files/GTseq_Genotyper_v3_Tetra.pl 103.1-genotyper-in.csv data/align/5D58_3_S160_L002_R1_001.fastq > outputs/102/5D58_3_S160.genos`

```{r}
test<-read_tsv("meta/test100.csv")
test<-test %>% mutate(Fastq=paste0("data/align/",Sample,"_L002_R1_001.fastq")) %>%
  mutate(Command=paste0("perl ./data-processing-files/GTseq_Genotyper_v3_Tetra.pl", " 103.1-genotyper-in.csv ",
         Fastq, " > ", "outputs/103/",Sample,".genos"))
test
```


cpan String::Approx
Installed in 
Installing /home/maccamp/perl5/lib/perl5/x86_64-linux-thread-multi/auto/String/Approx/Approx.so
Installing /home/maccamp/perl5/lib/perl5/x86_64-linux-thread-multi/String/Approx.pm
Installing /home/maccamp/perl5/man/man3/String::Approx.3

For me, I have perl installed with conda, so also maybe

conda install -c bioconda perl-string-approx


But, skipping the interpreter (/usr/bin/perl) works


perl ./data-processing-files/GTseq_Genotyper_v3_Tetra.pl 103.1-genotyper-in.csv data/align/5D58_3_S160_L002_R1_001.fastq > outputs/103/5D58_3_S160.genos
cata/align/5D58_3_S160_L002_R1_001.fastq,Raw-Reads:141406,On-Target reads:1634,%On-Target:1.2,IFI_score:9.24
Atr_30916-30,T=160,C=28,5.714,0000,NA,0,0     ,188,97.9,11.506,Pass,3
Atr_30916-62,G=0,A=0,1.000,0000,NA,0,0     ,0,0.0,0.000,Pass,0
Atr_39023-30,G=171,A=10,17.100,0000,NA,0,0    ,181,93.8,11.077,Pass,9
Atr_39023-56,C=0,T=0,1.000,0000,NA,0,0     ,0,0.0,0.000,Pass,0
Atr_39071-37,G=108,A=26,4.154,GGGA,HET1,0,0     ,134,93.7,8.201,Pass,5
Atr_39071-73,G=0,A=0,1.000,0000,NA,0,0     ,0,0.0,0.000,Pass,0
Atr_45890-42,A=75,C=95,0.789,AACC,HET2,1.6,0 ,284,35.1,17.381,Pass,53
Atr_45890-59,G=0,T=0,1.000,0000,NA,1.6,0,0,0.0,0.000,Pass,0
Atr_46229-48,G=59,A=182,0.324,GAAA,HET3,0,0       ,241,97.2,14.749,Pass,3
Atr_46229-66,G=0,A=0,1.000,0000,NA,0,0    ,0,0.0,0.000,Pass,0
Atr_54038-28,A=91,G=1,91.000,AAAA,A1HOM,0,0     ,92,52.0,5.630,Pass,16
Atr_54038-48,A=0,G=0,1.000,0000,NA,0,0      ,0,0.0,0.000,Pass,0
Atr_75140-52,C=61,T=22,2.773,CCCT,HET1,0,0     ,83,100.0,5.080,Pass,0
Atr_75140-58,C=0,T=0,1.000,0000,NA,0,0      ,0,0.0,0.000,Pass,0
Atr_76571-21,C=104,A=327,0.318,CAAA,HET3,0,0,431,72.0,26.377,Pass,6
Atr_76571-26,C=0,A=0,1.000,0000,NA,0,0       ,0,0.0,0.000,Pass,0

perl ./data-processing-files/GTseq_Genotyper_v3_Tetra.pl data-processing-files/genotyper_input.csv data/align/5D58_3_S160_L002_R1_001.fastq > outputs/103/5D58_3_S160-org.genos

base) maccamp@farm:~/kootenai$ grep Atr_30916-62 outputs/103/5D58_3_S160-org.genos 
Atr_30916-62,G=182,A=0,1820.000,GGGG,A1HOM,0,0,182,94.8,0.386,Pass,9

Seems fine.  

perl ./data-processing-files/GTseq_Genotyper_v3_Tetra.pl data-processing-files/genotyper_input-mac.csv data/align/5D58_3_S160_L002_R1_001.fastq > outputs/103/5D58_3_S160-alt.genos




Atr_30916-62,G=182,A=0,1820.000,GGGG,A1HOM,0,0,182,94.8,0.386,Pass,9

