---
title: "104-primer-mismatch"
output: html_document
date: "2023-09-22"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

```{r}
library(tidyverse)
```


This is a placeholder to investigate mismatches with the primer sequence leading to low input data.

```{r}
orgs<-read_csv("outputs/103/100-org-combined.csv",
                col_names=c("Locus","G1","G2","X4","Genotype","Call","X7","X8","X9","X10","X11","X12","X13"))
orgs
```


KT10_JV_01_S1514_L002_R1_001.fastq,Raw-Reads:59582,On-Target reads:10173,%On-Target:17.1,IFI_score:1.76

What may have a primer mismatch....

Need to identify variable sites in first 20 bp of seqs. That shoud be easy.

We have 405 common variable sites.  
 gunzip -c snps-90-all-seqs.geno.gz | wc
    405  640710 1950545

(base) maccamp@farm:~/kootenai/outputs/102$ gunzip -c snps-90-all-seqs.geno.gz | cut -f 1-2 | sort -n -k 2 | head -n 20
Atr_54160-70,T,C,TCATTTTTGGGTC,TCATTTTCGGGTC,TTTGCGGTTCAGATGTGGGT,0,0	9
Atr_64006-52,C,T,GGTTTACCCTGAAG,GGTTTACTCTGAAG,CTGGATGGGGAGCGGTTTAC,0,0	9
Atr_70167-40,C,T,CAGACCTCGTCCTG,CAGACCTTGTCCTG,GCATGCATCAGATAGCCAGC,0,0	12
Atr_48754-60,G,A,CTGCAACGACGTGT,CTGCAACAACGTGT,GCTTTCTCCTTTCAGTGGCC,0,0	14
Atr_64006-52,C,T,GGTTTACCCTGAAG,GGTTTACTCTGAAG,CTGGATGGGGAGCGGTTTAC,0,0	14
Atr_1720-33,C,A,GGGGGCTTTGAGAGC,GGGGGATTTGAGAGC,GGCAAAGCAGAAGGGGG,0,0.7	18
Atr_54160-70,T,C,TCATTTTTGGGTC,TCATTTTCGGGTC,TTTGCGGTTCAGATGTGGGT,0,0	18
Atr_29086-33,C,T,TTGTCAACGGGTTT,TTGTCAATGGGTTT,GGTTTGGTTCAACCGCACAG,0,0	19
Atr_58434-42,G,A,CCAGTCCGTTTTCC,CCAGTCCATTTTCC,GCAGAACCACGCCAGTCC,0,1.5	19
Atr_70167-40,C,T,CAGACCTCGTCCTG,CAGACCTTGTCCTG,GCATGCATCAGATAGCCAGC,0,0	19
Atr_18334-74,A,G,ATGGTGGAC,ATGGTGGGC,GGAAGTTGGTCCTCAGAGGC,0,0	20
Atr_23070-31,C,A,CAGTCTCCGCATTT,CAGTCTCAGCATTT,GCCTTTGCCCTCCAGTCTC,0,0	20
Atr_36485-28,C,T,CCTCATCCTAGCCC,CCTCATCTTAGCCC,GGTGCTAGCAGCCCTCATC,0,0	20
Atr_60368-72,A,T,AGGAACCAGCA,AGGAACCTGCA,CGATGGCCTCCAAGGAACC,0,0	20
Atr_10428-71,T,A,AGAGAGCTAGCA,AGAGAGCAAGCA,CGCACACACACAAAGAGAGC,0,2	21
Atr_11864-68,T,C,ATATCGCTGACCAC,ATATCGCCGACCAC,GGTACCACCGCCAATATCGC,0,0	21
Atr_14917-56,C,T,AACACATCTTAACA,AACACATTTTAACA,TGCGCCACCTAGAAACACAT,0,0	21
Atr_18295-67,T,C,ACCTGCATGGCAAA,ACCTGCACGGCAAA,CTCGTCTTGAGGCACCTGC,0,0	21
Atr_30271-58,G,A,AGGAGTCGTATTTC,AGGAGTCATATTTC,ATGCAGAATCCACAGGAGTC,0,0	21
Atr_40110-57,G,A,TAGGGCCGCATGCA,TAGGGCCACATGCA,GCACTGTGCCACTGACATAT,0,0	21

Would expect Atr_54160-70 to be a poor performer. Let's investigate more.

```{r}
mismatches<-filter(orgs, Locus %in% c("Atr_54160-70","Atr_64006-52","Atr_70167-40","Atr_48754-60"))
```

```{r}
summary<-mismatches %>% group_by(Locus, Call) %>% summarize(Count=n())
summary
```

NAs aren't counted, so something like this:
```{r}
ggplot(summary) + geom_col(aes(x=Locus, y=Count, fill=Call)) +
  theme_bw() +
  theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1)) +
  theme(panel.grid = element_blank()) +
  scale_fill_viridis_d(option="H", na.value = "grey")
```

Atr_70167-40 has a lot of missing data.

I wonder if I can use (C|T) or whatever in the primer string.  That is site tweleve would be:

GCATGCATCAGATAGCCAGC
GCATGCATCAG(A|T)TAGCCAGC # doesn't work
GCATGCATCAG # doesn't work
`GCATGCATCAG\wTAGCCAGC` doesn't work
GCATGCATCAG.TAGCCAGC
        $F_Primer{$info[0]} = substr $info[5], 0, 14; $ this seems to say it takes the first 15 characters of the F_Primer


(base) maccamp@farm:~/kootenai/data/new-ref$ grep GCATGCATCAG --color test.fasta shows a polymorphic site in the primer.

Atr_48754-60 has a lot of missing data 

GCTTTCTCCTTTCAGTGGCC
GCTTTCTCCTTTC(A|T)GTGGCC

Something to try!   

Let's clean up our genotype file and get MAFs. Maybe the header is messing things up? let's try to clean up the naming in original ref and call snps again.
/home/maccamp/kootenai/data/align-new-ref-clean
/home/maccamp/kootenai/new-ref/new-ref-short-names.fasta

Missing an alignment, redoing, now calling. 

```{sh, eval=FALSE}
srun -t 12:00:00 -p bigmemh --mem=128G --nodes=1 --ntasks-per-node=1 --cpus-per-task=12 $HOME/angsd/angsd -nthreads 24 \
-bam bamlists/aft01-clean.bamlist -minInd 1422 \
-out $HOME/kootenai/outputs/104/snps-90-all-seqs  \
-minMaf 0.10 -minMapQ 20 -minQ 20 -GL 1 -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-doGeno 4 -doPost 1 -postCutoff 0.95 -doPlink 2  > outputs/104/snps-90-all-seqs.out 2> outputs/104/snps-90-all-seqs.err &


plink --tped snps-90-all-seqs.tped --tfam snps-90-all-seqs.tfam --out snps-90-all-seqs.binary --recode --allow-extra-chr --noweb
plink --ped snps-90-all-seqs.binary.ped --map snps-90-all-seqs.binary.map --recode vcf --allow-extra-chr -out snps-90-all-seqs

bcftools +fill-tags snps-90-all-seqs.vcf -- -t MAF  > snps-90-all-seqs-maf.vcf

```




I could regex in a new primer seq for all fastqs...

Atr_54160-70,T,C,TCATTTTTGGGTC,TCATTTTCGGGTC,TTTGCGGTTCAGATGTGGGT,0,0	9

Expect sites 9 (G/T) and 18 (G/A) to be polymorphic. Not seeing T/C polymorphism in this seq called, but they are in the fastq. Shrug emoji.   

TTTGCGGTTCAGATGTGGGT
TTTGCGGTTCAGATGTGGGT CACACAGATCGGAAGAGCACACGTCTGAACTCCAGTCACGGCTACATCTCGTATGCCGTCTTCTGCTTGAAAAAAAAAAAN
TTTGCGGTGCAGATGTGAGT CACACAGATCGGAAGAGCACACGTCTGAACTCCAGTCACGGCTACATCTCGTATGCCGTCTTCTGCTTGAAAAAAAAAAAA


 grep "^TTTGCGGTTCAGATGTGGGT" 3A21_7_S84_L002_R1_001.fastq | wc -l
7597
(base) maccamp@farm:~/kootenai/outputs/104$ grep "^TTTGCGGTGCAGATGTGAGT" 3A21_7_S84_L002_R1_001.fastq | wc -l
6868

(base) maccamp@farm:~/kootenai/outputs/104$ grep "^TTTGCGGT" 3A21_7_S84_L002_R1_001.fastq | wc -l
14728

7597+6868

Pretty darn close.
Are these linked smps? So something like perl -pi -e 's/^TTTGCGGTGCAGATGTGAGT/TTTGCGGTTCAGATGTGGGT/
Let's try a trial individual:


(base) maccamp@farm:~/kootenai/outputs/104$ cp ../../data/align/3A21_7_S84_L002_R1_001.fastq .

perl -pi -e 's/^TTTGCGGTGCAGATGTGAGT/TTTGCGGTTCAGATGTGGGT/' 3A21_7_S84_L002_R1_001.fastq

(base) maccamp@farm:~/kootenai/outputs/104$ grep "^TTTGCGGTTCAGATGTGGGT" 3A21_7_S84_L002_R1_001.fastq | wc -l
14465

  perl -pi -e 's/^TTTGCGGT[G|T]CAGATGTG[A|G]GT/TTTGCGGTTCAGATGTGGGT/' 3A21_7_S84_L002_R1_001.fastq
  
  grep "^TTTGCGGTTCAGATGTGGGT"  3A21_7_S84_L002_R1_001.fastq | wc -l
  14539


grep "^TTTGCGGTGCAGATGTGAGT" 3A21_7_S84_L002_R1_001.fastq | wc -l
0

--
Atr_64006-52

positions 9/14 polymorphic

grep ^CTGGATGGGGAGCGGTTTAC 3A21_7_S84_L002_R1_001.fastq  | wc -l
2055
grep ^CTGGATGGAGAGCAGTTTAC 3A21_7_S84_L002_R1_001.fastq  | wc -l
743


perl -pi -e 's/^CTGGATGG[A|G]GAGC[A|G]GTTTAC/CTGGATGGGGAGCGGTTTAC/' 3A21_7_S84_L002_R1_001.fastq
(base) maccamp@farm:~/kootenai/outputs/104$ grep ^CTGGATGGGGAGCGGTTTAC 3A21_7_S84_L002_R1_001.fastq  | wc -l
2880

--
Atr_70167-40,
sites 12/19 polymorphic

GCATGCATCAGATAGCCAAC
GCATGCATCAGATAGCCAGC
GCATGCATCAGTTAGCCACC
 TAGCCA[A|G|C]C
       GCATGCATCAGATAGCCAGC
 grep ^GCATGCATCAGATAGCCAGC 3A21_7_S84_L002_R1_001.fastq | wc -l
431

perl -pi -e 's/^GCATGCATCAG[A|T]TAGCCA[A|G|C]C/GCATGCATCAGATAGCCAGC/' 3A21_7_S84_L002_R1_001.fastq

grep ^GCATGCATCAGATAGCCAGC 3A21_7_S84_L002_R1_001.fastq | wc -l
1880

--
Atr_48754-60

GCTTTCTCCTTTCAGTGGCC
GCTTTCTCCTTTCAGTGGCC
GCTTTCTCCTTTCTGTGGCC

grep "^GCTTTCTCCTTTCTGTGGCC" 3A21_7_S84_L002_R1_001.fastq | wc -l 
50



pi -e 's/^GCTTTCTCCTTTC[A|T]GTGGCC/GCTTTCTCCTTTCAGTGGCC/' 3A21_7_S84_L002_R1_001.fastq

(base) maccamp@farm:~/kootenai/outputs/104$ grep "^GCTTTCTCCTTTCTGTGGCC" 3A21_7_S84_L002_R1_001.fastq | wc -l
0
(base) maccamp@farm:~/kootenai/outputs/104$ grep "^GCTTTCTCCTTTCAGTGGCC" 3A21_7_S84_L002_R1_001.fastq | wc -l 
125

low numbers here     

---      
Atr_64006-52
Already did!

## strings
Atr_54160-70,T,C,TCATTTTTGGGTC,TCATTTTCGGGTC,TTTGCGGTTCAGATGTGGGT,0,0	9
Atr_64006-52,C,T,GGTTTACCCTGAAG,GGTTTACTCTGAAG,CTGGATGGGGAGCGGTTTAC,0,0	9
Atr_70167-40,C,T,CAGACCTCGTCCTG,CAGACCTTGTCCTG,GCATGCATCAGATAGCCAGC,0,0	12
Atr_48754-60,G,A,CTGCAACGACGTGT,CTGCAACAACGTGT,GCTTTCTCCTTTCAGTGGCC,0,0	14
Atr_64006-52,C,T,GGTTTACCCTGAAG,GGTTTACTCTGAAG,CTGGATGGGGAGCGGTTTAC,0,0	14

                
perl -pi -e 's/^TTTGCGGT[G|T]CAGATGTG[A|G]GT/TTTGCGGTTCAGATGTGGGT/' 3A21_7_S84_L002_R1_001.fastq
perl -pi -e 's/^CTGGATGG[A|G]GAGC[A|G]GTTTAC/CTGGATGGGGAGCGGTTTAC/' 3A21_7_S84_L002_R1_001.fastq
perl -pi -e 's/^GCATGCATCAG[A|T]TAGCCA[A|G|C]C/GCATGCATCAGATAGCCAGC/' 3A21_7_S84_L002_R1_001.fastq
perl -pi -e 's/^GCTTTCTCCTTTC[A|T]GTGGCC/GCTTTCTCCTTTCAGTGGCC/' 3A21_7_S84_L002_R1_001.fastq
