---
title: "201-call-broodstock"
output: html_document
date: "2023-10-12"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

```{r}
library(tidyverse)
```

Call genotypes for the parents. Have bamlist;

```{r}
parents<-read_csv("meta/kootenai-all.csv") %>% filter(Type=="broodstock")
parents$Sample<-gsub("/home/maccamp/kootenai/data/align/","",parents$Sample)
parents
```

346 parents to get some genos

```{r}
parents <- parents %>% 
  mutate(Command=paste0("perl ./data-processing-files/GTseq_Genotyper_v3_Tetra.pl", 
                         " data-processing-files/genotyper_input.csv ",
         paste0("data/align/",SeqID,"_L002_R1_001.fastq"), " > ", "outputs/202/",Sample,".org"))

parents %>% select(Command) %>% write_tsv("201.1-broodstock-commands.sh", col_names = FALSE)
```

module load parallel    
srun -p high -t 2:00:00 --nodes=1 --mem=32GB parallel -j 12 < 201.1-broodstock-commands.sh     

Reading in genos


```{r}
readFile<-function(file) {
  df<-read_csv(file, skip=1, col_names=c("Locus","G1","G2","X4","Genotype","Call","X7","X8","X9","X10","X11","X12","X13"))
  m<-read.table(file, nrows = 1) %>% as_tibble() %>% separate(V1, into=c("Seq","Reads","T"), sep=",") %>% 
    separate(V2, into=c("AlignedReads","Percent","IFI"), sep=",")
  df<-df %>% bind_cols(m)
  df$SeqID<-gsub("_L002_R1_001.fastq","",df$Seq)
  df$Sample<-gsub("_S\\d+$","",df$SeqID)
  return(df)
}
```

```{r, eval=FALSE}
files<-list.files(path="outputs/202", pattern="*org", full.names= TRUE)
targets<-bamall %>% mutate(File=paste0("outputs/202/", Sample,".org"))
t2<-targets %>% filter(File %in% files) 
t2 %>% group_by(Type) %>% summarize(Count=n())
```


```{r, eval=FALSE}
reads<-lapply(t2$File, readFile)
reads<-plyr::rbind.fill(reads)
reads<-as_tibble(reads)
save(reads, file="outputs/202/reads.rda")
```